File: admin/app/admin/logic/admin_menu.go
package logic

import (
	"admin/app/admin/dao"
	"admin/app/common"
	"admin/app/gen/model"
	"admin/code"
	"admin/pkg/utils"
	"admin/proto/admin_proto"
	"admin/proto/code_proto"
	"context"
	"errors"
	"github.com/gin-gonic/gin"
	"gorm.io/gorm"
)

type AdminMenuLogic struct {
	db *dao.AdminMenu
}

func NewAdminMenuLogic() *AdminMenuLogic {
	return &AdminMenuLogic{
		db: adminMenuDao,
	}
}

func (a *AdminMenuLogic) List(ctx *gin.Context, params *admin_proto.MenuListReq) (data *admin_proto.MenuListRespData, err error) {
	total, rows, err := a.db.FindList(ctx, params)
	if err != nil {
		return nil, err
	}
	data = &admin_proto.MenuListRespData{}
	data.Total = total
	data.Rows, err = a.HandleListData(rows)
	return data, err
}

func (a *AdminMenuLogic) Tree(ctx *gin.Context) ([]*admin_proto.MenuTreeItem, error) {
	allMenus, err := a.db.FindAll(ctx)
	if err != nil {
		return nil, err
	}
	menusMap := make(map[int32][]*admin_proto.MenuTreeItem)
	for _, item := range allMenus {
		if _, ok := menusMap[item.ParentID]; !ok {
			menusMap[item.ParentID] = make([]*admin_proto.MenuTreeItem, 0)
		}
		tmp := &admin_proto.MenuTreeItem{}
		_ = utils.BeanCopy(tmp, item)
		tmp.Id = item.ID
		tmp.ParentId = item.ParentID
		tmp.Enabled = item.IsEnabled
		tmp.HideInMenu = item.IsHideInMenu
		tmp.HideChildrenInMenu = item.IsHideChildrenInMenu
		tmp.CreateTime = item.CreatedAt.Unix()
		tmp.ModifyTime = item.UpdatedAt.Unix()
		tmp.Children = make([]*admin_proto.MenuTreeItem, 0)
		menusMap[item.ParentID] = append(menusMap[item.ParentID], tmp)
	}
	data := common.MenuTree(menusMap, nil, 0, 1, 4)
	return common.GetMenuTreeWithTop(data), nil
}

func (a *AdminMenuLogic) Add(ctx *gin.Context, params *admin_proto.MenuAddReq) error {
	data := &model.AdminMenu{
		ParentID:             params.ParentId,
		Path:                 params.Path,
		Name:                 params.Name,
		Key:                  params.Key,
		Describe:             params.Describe,
		Icon:                 params.Icon,
		Sort:                 params.Sort,
		Redirect:             params.Redirect,
		Component:            params.Component,
		IsHideInMenu:         params.HideInMenu,
		IsHideChildrenInMenu: params.HideChildrenInMenu,
		IsEnabled:            params.Enabled,
	}
	return a.db.Create(ctx, data)
}

func (a *AdminMenuLogic) Info(ctx *gin.Context, params *admin_proto.MenuInfoReq) (*admin_proto.MenuItem, error) {
	data, err := a.db.FindById(ctx, params.MenuId)
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, code.NewCodeError(code_proto.ErrorCode_RecordNotExist, err)
		}
		return nil, err
	}
	return a.HandleItemData(data)
}

func (a *AdminMenuLogic) Edit(ctx *gin.Context, params *admin_proto.MenuEditReq) error {
	info, err := a.db.FindById(ctx, params.MenuId)
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return code.NewCodeError(code_proto.ErrorCode_RecordNotExist, err)
		}
		return err
	}
	info.ParentID = params.ParentId
	info.Path = params.Path
	info.Name = params.Name
	info.Key = params.Key
	info.Describe = params.Describe
	info.Icon = params.Icon
	info.Sort = params.Sort
	info.Redirect = params.Redirect
	info.IsHideInMenu = params.HideInMenu
	info.IsHideChildrenInMenu = params.HideChildrenInMenu
	info.IsEnabled = params.Enabled
	return a.db.Update(ctx, info)
}

func (a *AdminMenuLogic) Enable(ctx *gin.Context, params *admin_proto.MenuEnableReq) error {
	info, err := a.db.FindById(ctx, params.MenuId)
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return code.NewCodeError(code_proto.ErrorCode_RecordNotExist, err)
		}
		return err
	}
	if info.IsEnabled == params.Enabled {
		return nil
	}
	return a.db.Enable(ctx, params.MenuId, params.Enabled)
}

func (a *AdminMenuLogic) Delete(ctx *gin.Context, params *admin_proto.MenuDeleteReq) error {
	info, err := a.db.FindById(ctx, params.MenuId)
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return code.NewCodeError(code_proto.ErrorCode_RecordNotExist, err)
		}
		return err
	}
	if info.IsEnabled {
		return code.NewCodeError(code_proto.ErrorCode_RecordNValidCanNotDeleted, nil)
	}
	return a.db.Delete(ctx, params.MenuId)
}

func (a *AdminMenuLogic) Permissions(ctx *gin.Context, params *admin_proto.MenuPermissionsReq) (*admin_proto.MenuPermissions, error) {
	info, err := a.db.FindById(ctx, params.MenuId)
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, code.NewCodeError(code_proto.ErrorCode_RecordNotExist, err)
		}
		return nil, err
	}
	data := &admin_proto.MenuPermissions{
		Menu:        &admin_proto.MenuTreeItem{},
		Permissions: make([]*admin_proto.PermissionApiItem, 0),
	}
	_ = utils.BeanCopy(data.Menu, info)
	permissions, err := adminPermissionDao.FindPermissionsByMenuId(ctx, info.ID)
	if err != nil {
		return nil, err
	}
	if len(permissions) == 0 {
		data.Permissions = common.AdminPermissionEnumList(info.ID, info.Key)
		return data, nil
	}
	// TODO
	permissionIdsMap := make(map[int32]struct{})
	permissionIds := make([]int32, 0, len(permissions))
	for _, item := range permissions {
		if _, ok := permissionIdsMap[item.ID]; !ok {
			permissionIdsMap[item.ID] = struct{}{}
			permissionIds = append(permissionIds, item.ID)
		}

	}
	apiList, err := adminAPIDao.FindByPermissionIds(ctx, permissionIds)
	if err != nil {
		return nil, err
	}
	apiMap := make(map[int32][]*admin_proto.ApiItem)
	for _, item := range apiList {
		if _, ok := apiMap[item.PermissionId]; !ok {
			apiMap[item.PermissionId] = make([]*admin_proto.ApiItem, 0, len(apiList))
		}
		apiMap[item.PermissionId] = append(apiMap[item.PermissionId], item)
	}
	for _, item := range data.Permissions {
		if val, ok := apiMap[item.Id]; ok {
			item.Apis = val
		}
	}
	return data, nil
}

func (a *AdminMenuLogic) HandleListData(list []*model.AdminMenu) (rows []*admin_proto.MenuItem, err error) {
	for _, item := range list {
		data, err := a.HandleItemData(item)
		if err != nil {
			return nil, err
		}
		rows = append(rows, data)
	}
	return rows, nil
}

func (a *AdminMenuLogic) HandleItemData(item *model.AdminMenu) (data *admin_proto.MenuItem, err error) {
	data = &admin_proto.MenuItem{}
	err = utils.BeanCopy(data, item)
	if err != nil {
		return nil, err
	}
	return data, nil
}

func (a *AdminMenuLogic) getMyMenusMap(ctx context.Context, pageIds []int32) ([]*admin_proto.MenuItem, error) {
	allMenus, err := a.db.FindAllValid(ctx)
	if err != nil {
		return nil, err
	}
	allMenusMap := make(map[int32]*model.AdminMenu)
	for _, item := range allMenus {
		allMenusMap[item.ID] = item
	}
	arr1 := getAllRelatedMenusByPageIds(allMenusMap, pageIds)
	arr2 := getAllChildrenPagesByPageIds(allMenusMap, pageIds)
	arr := append(arr1, arr2...)
	menus := make([]*admin_proto.MenuItem, 0, len(arr))
	m := make(map[string]struct{}, len(arr))
	for _, item := range arr {
		if _, ok := m[item.Key]; !ok {
			m[item.Key] = struct{}{}
			menus = append(menus, item)
		}
	}
	return menus, nil
}

func getAllRelatedMenusByPageIds(menusMap map[int32]*model.AdminMenu, pageIds []int32) (data []*admin_proto.MenuItem) {
	for _, pageId := range pageIds {
		arr := getAllFatherMenuByChildrenId(menusMap, pageId)
		if len(arr) > 0 {
			data = append(data, arr...)
		}
	}
	return data
}

func getAllChildrenPagesByPageIds(menusMap map[int32]*model.AdminMenu, pageIds []int32) (data []*admin_proto.MenuItem) {
	for _, pageId := range pageIds {
		arr := getAllChildrenMenuByChildrenId(menusMap, pageId)
		if len(arr) > 0 {
			data = append(data, arr...)
		}
	}
	return data
}

func getAllFatherMenuByChildrenId(menusMap map[int32]*model.AdminMenu, parentId int32) (data []*admin_proto.MenuItem) {
	if menu, ok := menusMap[parentId]; ok {
		data = append(data, &admin_proto.MenuItem{
			Key:       menu.Key,
			Path:      menu.Path,
			Name:      menu.Name,
			Icon:      menu.Icon,
			Component: menu.Component,
		})
		arr := getAllFatherMenuByChildrenId(menusMap, menu.ParentID)
		if len(arr) > 0 {
			data = append(data, arr...)
		}
	}
	return data
}

func getAllChildrenMenuByChildrenId(menusMap map[int32]*model.AdminMenu, parentId int32) (data []*admin_proto.MenuItem) {
	for menuId, menu := range menusMap {
		if menu.ParentID == parentId {
			data = append(data, &admin_proto.MenuItem{
				Key:       menu.Key,
				Path:      menu.Path,
				Name:      menu.Name,
				Icon:      menu.Icon,
				Component: menu.Component,
			})

			tmpIds := make(map[int32]*model.AdminMenu)
			for _, item := range menusMap {
				if item.ParentID != parentId {
					tmpIds[item.ID] = item
				}
			}
			arr := getAllChildrenMenuByChildrenId(menusMap, menuId)
			if len(arr) > 0 {
				data = append(data, arr...)
			}
		}
	}
	return data
}

File: admin-vue3x/src/types/admin_menu.ts
import type { ReqListBase } from './common';

/** 菜单列表请求 */
export interface ReqAdminMenuList extends ReqListBase {
  key?: string;
  path?: string;
  name?: string;
  parentId?: number;
}

/** 创建/修改菜单 */
export interface ReqAdminMenuAdd {
  id?: number;
  parentId?: number;
  path?: string;
  name: string;
  key: string;
  describe?: string;
  icon?: string;
  sort?: number;
  redirect?: string;
  component?: string;
  hideChildrenInMenu?: boolean;
  hideInMenu?: boolean;
  enabled?: boolean;
  /** 表单使用的字段：是否隐藏菜单 */
  isHideInMenu?: boolean;
  /** 表单使用的字段：是否隐藏子菜单 */
  isHideChildrenInMenu?: boolean;
  /** 表单使用的字段：是否启用 */
  isEnabled?: boolean;
}

/** 启用禁用菜单 */
export interface ReqAdminMenuEnable {
  menuId?: number;
  enabled?: boolean;
}

/** 是否显示菜单 */
export interface ReqAdminMenuShow {
  menuId?: number;
  show?: boolean;
}

/** 删除菜单 */
export interface ReqAdminMenuDelete {
  menuId?: number;
}

/** 菜单详情 */
export interface ReqAdminMenuInfo {
  menuId?: number;
}

/** 菜单权限列表 */
export interface ReqAdminMenuPermissions {
  menuId?: number;
}

/** 页面菜单列表 */
export interface ReqAdminMenuPages {
  all?: boolean;
}

/** 页面模块权限列表 */
export interface ReqAdminMenuMode {}

/** 菜单 */
export interface MenuItem {
  /** 菜单唯一键 */
  key?: string;
  /** 菜单路由 */
  path?: string;
  /** 菜单名称 */
  name?: string;
  /** 菜单图标 */
  icon?: string;
  component?: string;
  authority?: string;
  hideInMenu?: boolean;
  hideChildrenInMenu?: boolean;
  routes?: MenuItem[];
}

/** 有效菜单树 */
export interface MenuTreeItem {
  /** 菜单层级 */
  level?: number;
  /** 菜单自增ID */
  id?: number;
  /** 菜单唯一键 */
  key?: string;
  /** 菜单名称 */
  name?: string;
  /** 父级菜单ID */
  parentId?: number;
  /** 菜单描述 */
  describe?: string;
  /** 菜单路径 */
  path?: string;
  /** 重定向地址 */
  redirect?: string;
  /** 组件名称 */
  component?: string;
  /** 菜单排序 */
  sort?: number;
  /** 菜单图标 */
  icon?: string;
  /** 是否在菜单中隐藏子菜单 */
  hideChildrenInMenu?: boolean;
  /** 是否隐藏菜单 */
  hideInMenu?: boolean;
  /** 是否启用 */
  enabled?: boolean;
  createTime?: number;
  modifyTime?: number;
  children?: MenuTreeItem[];
}

/** 菜单模型 */
export interface AdminMenuModel {
  id?: number;
  parentId?: number;
  path?: string;
  name?: string;
  key?: string;
  describe?: string;
  icon?: string;
  sort?: number;
  redirect?: string;
  component?: string;
  isHideInMenu?: boolean;
  isHideChildrenInMenu?: boolean;
  isEnabled?: boolean;
  createdAt?: string;
  updatedAt?: string;
}

/** 菜单模块 */
export interface MenuModeItem {
  /** 模块菜单ID */
  modelId?: number;
  /** 模块菜单名称 */
  modelName?: string;
  /** 模块下面的页面列表 */
  pages?: MenuPageItem[];
}

/** 模块页面 */
export interface MenuPageItem {
  /** 页面菜单ID */
  pageId?: number;
  /** 页面菜单名称 */
  pageName?: string;
  /** 页面菜单权限 */
  permissions?: MenuPagePermissions[];
}

/** 模块页面权限 */
export interface MenuPagePermissions {
  /** 权限ID */
  permissionId?: number;
  /** 权限名称 */
  permissionName?: string;
  /** 权限类型 */
  permissionType?: string;
  /** 权限类型名称 */
  permissionTypeName?: string;
}

/** 菜单权限 */
export interface MenuPermissionItem {
  id?: number;
  name?: string;
  key?: string;
  type?: string;
  typeName?: string;
  describe?: string;
  enabled?: boolean;
  menuId?: number;
}
